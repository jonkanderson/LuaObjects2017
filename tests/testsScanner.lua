local args = {...}
local CL, FN, COL, S = dofile('../root.lua')

local Pm = FN.newPathManager()
Pm:setArgs(args)
Pm:setHere('.')
Pm:setModDir('../mods')

Pm:loadMod('modScanner.lua')

--====================================================== main code
-- Good for debugging.
--FN.reportAllKeys()

local sc = CL.Scanner:newInstance()
local tokens

local text = [[0293893 hello900   'this is '' a string'
		begin some stuff begin other end end]]

sc:setStreamFromString(text)

function sc:nextToken()
	local token = nil

	token = self:ifNextMatchScan("begin", function(newToken, state)
		if state == "start" then
			local t = self:manyWhitespaces()
			if not t then return "abort" end
			newToken:addSubtoken(t)
			return "continue", "content"
		elseif state == "content" then
			if self:testMatch("end") then
				return "continue", "final"
			else
				t = self:nextToken()
				if not t then return "abort" end
				newToken:addSubtoken(t)
			end
			return "continue", "content"
		elseif state == "final" then
			t = self:nextToken()
			newToken:addSubtoken(t)
			return "final", nil
		else
			error("Unhandled state: "..tostring(state))
		end
	end)
	if token then return token:label("BeginEndChunk") end

	token = self:manyWhitespaces()
	if token then return token:label("Whitespace") end

	token = self:manyDigits()
	if token then return token:label("Digits") end

	token = self:manyLetters()
	if token then return token:label("Letters") end

	token = self:ifNextMatchScan("'", function()
		if self:nextMatch("''") then return "continue" end
		if self:nextMatch("'") then return "final" end
		if self:atEnd() then return "abort" end
		self:nextAny()
		return "continue"
	end)
	if token then return token:label("String") end

	error("No Token recognized.")
end

print("Parsing Text:")
print(">>"..text.."<<")

tokens = sc:scanAllTokens()
for _,t in ipairs(tokens) do
	print(t:debugString())
end
assert(#tokens == 8)
