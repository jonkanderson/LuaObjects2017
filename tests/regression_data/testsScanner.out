Parsing Text:
>>0293893 hello900   'this is '' a string'
		begin some stuff begin other end end<<
[[0293893]](Digits, 1, 7)
[[ ]](Whitespace, 8, 8)
[[hello]](Letters, 9, 13)
[[900]](Digits, 14, 16)
[[   ]](Whitespace, 17, 19)
[['this is '' a string']](String, 20, 40)([[']](Start, 20, 20))
[[
		]](Whitespace, 41, 43)
[[begin some stuff begin other end end]](BeginEndChunk, 44, 79)([[begin]](Start, 44, 48)[[ ]](nil, 49, 49)[[some]](Letters, 50, 53)[[ ]](Whitespace, 54, 54)[[stuff]](Letters, 55, 59)[[ ]](Whitespace, 60, 60)[[begin other end]](BeginEndChunk, 61, 75)([[begin]](Start, 61, 65)[[ ]](nil, 66, 66)[[other]](Letters, 67, 71)[[ ]](Whitespace, 72, 72)[[end]](Letters, 73, 75))[[ ]](Whitespace, 76, 76)[[end]](Letters, 77, 79))
